smop/libscripts/octave-4.0.2/doc/interpreter/octave-config.1
.\" Copyright (C) 2003-2015 Dirk Eddelbuettel
.\"
.\" This file is part of Octave.
.\"
.\" Octave is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 3 of the License, or (at
.\" your option) any later version.
.\"
.\" Octave is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with Octave; see the file COPYING.  If not, see
.\" <http://www.gnu.org/licenses/>.
.\"
.\" This page was contributed by Dirk Eddelbuettel <edd@debian.org>.
.\" This page was completely re-written by Rik <octave@nomad.inbox5.com>
.\" --------------------------------------------------------------------
.de Vb \" (V)erbatim (b)egin.  Use fixed width font and no justification
.ft CW
.nf
..
.de Ve \" (V)erbatim (e)nd.  Return to regular font and justification
.ft R
.fi
..
.\" --------------------------------------------------------------------
.TH OCTAVE-CONFIG 1 "3 January 2014" "GNU Octave"
.SH NAME
octave-config - GNU Octave component and library information retrieval
.SH SYNOPSIS
\fBoctave-config\fP [\fIOPTION\fP]...
.SH DESCRIPTION
.B octave-config
is a tool to obtain directory information for \fB.oct\fP and \fB.m\fP files for
.BR octave (1).
.SH OPTIONS
.TP 8
.B \-\-m-site-dir
Display the main directory for local, or site-specific, \fB.m\fP script files.
.TP
.B \-\-oct-site-dir
Display the main directory for local, or site-specific, \fB.oct\fP dynamic-link libraries.
.TP
.B \-p \fIVAR\fP\fR,\fB \-\-print \fIVAR\fP
Print the default value of the Octave configuration variable \fIVAR\fP.
Recognized variables are:
.RS
.Vb
   API_VERSION             LOCALAPIOCTFILEDIR
   ARCHLIBDIR              LOCALARCHLIBDIR
   BINDIR                  LOCALFCNFILEDIR
   CANONICAL_HOST_TYPE     LOCALOCTFILEDIR
   DATADIR                 LOCALSTARTUPFILEDIR
   DATAROOTDIR             LOCALVERARCHLIBDIR
   DEFAULT_PAGER           LOCALVERFCNFILEDIR
   EXEC_PREFIX             LOCALVEROCTFILEDIR
   FCNFILEDIR              MAN1DIR
   IMAGEDIR                MAN1EXT
   INCLUDEDIR              MANDIR
   INFODIR                 OCTFILEDIR
   INFOFILE                OCTINCLUDEDIR
   LIBDIR                  OCTLIBDIR
   LIBEXECDIR              PREFIX
   LOCALAPIARCHLIBDIR      STARTUPFILEDIR
   LOCALAPIFCNFILEDIR      VERSION
.Ve
.RE
.TP
.B \-v\fR,\fB \-\-version
Display the version number of
.BR octave (1).
.TP
\fB\-h\fP, \fB\-?\fP, \fB\-\-help\fP
Print help message.
.SH AUTHOR
John W. Eaton <jwe@octave.org>

This manual page was contributed by Dirk Eddelbuettel <edd@debian.org>
for the Debian GNU/Linux distribution but may be used by others.
.SH "SEE ALSO"
octave (1).


NEWS.rst
=============
Release notes
=============

    Version 0.33 is out

    New option -c --code 
        Quickly check how a particular statement
        is translated. Option -c implies -o- .

    Renamed option -D to -d (--debug).

    New regression count:

        $ cd libscripts && make regresss

    Currently 14 regressions

        
November 18, 2016
    Version 0.31 is out

    Changed naming of output files.
        If -o not specified, compiling foo.m
	now results in foo.py

    Using octave script library for testing
        $ wget ftp://ftp.gnu.org/ftp/octave/octave-4.0.2.tar.gz
	$ smop -a octave-4.0.2.tar.gz -g '*/scripts/*.m'
	$ ls -1 *.py | wc
	$ python -m py_compile *.py
	$ ls -1 *.pyc | wc

    New option -a --archive allows to get data from
        tar file, without extracting the contents
	to a temporary place.

    New option -g --glob-pattern allows to limit
        the amount of input data

    Removed option --ignore-errors.
        All errors are ignored now.

    
August 27, 2016
    Version 0.29 is out

    Test suite for regression testing is ready,
        with about 1000 m-scripts.  Scripts are translated
        to python, then the resulting py-files are loaded,
        but not yet run.
    
August 11, 2016
    Version 0.28 is out. 

    Line numbering information is included in the output.
        Enabled by default.  Disabled with ``--no-numbers``.

    Block comments are preserved now.
        Lines, containing anything but a comment, are
        preserved. Enabled by default.
        Disabled with ``--no-comments``.

    New command-line options
        ``--no-resolve,--no-backend`` useful mostly for
        debugging. New shortcuts -R -B -T -C -N
        
    Special ``%!`` comments are partially supported now.
        They are mostly useful in testing, and
        are used by Octave library and test suite. Disabled
        by default. Enabled with ``--testing-mode``.
        

June 19,2016
   After a year-long vacation I am back to active development.
   My first goal is adopting the Octave runtime and test suite.

October 23, 2014
   Downloaded ``mybench`` -- a collection of 20 or so
   micro-benchmarks originally meant to compare matlab and
   octave performance.  After succesfully running the first nine,
   the geometric mean of the speedup is 0.36,  which is cool.

    ==   ========   ======    ===========    =======
    //   name       octave    smop           speedup
    ==   ========   ======    ===========    =======
    1    rand       2.58      0.36           0.14
    2    randn      2.26      1.04           0.46
    3    primes     0.35      0.17           0.49
    4    fft2       2.75      1.13           0.41
    5    square     4.24      0              
    6    inv        4.38      2.26           0.53
    7    eig        17.95     9.09           0.51
    8    qr         3.06      1.83           0.60
    9    shur       5.98      2.31           0.39
    10   roots      8.31      2.02           0.24
    ==   ========   ======    ===========    =======

October 15, 2014
   Version 0.26.3 is available for beta testing.
   Next version 0.27 is planned to compile octave
   ``scripts`` library, which contains over 120 KLOC in
   almost 1,000 matlab files. There  are 13 compilation
   errors with smop 0.26.3 .



smop/libscripts/octave-4.0.2/src/octave-config.in.cc
// %NO_EDIT_WARNING%
/*

Copyright (C) 2008-2015 Michael Goffioul

This file is part of Octave.

Octave is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 3 of the License, or (at your
option) any later version.

Octave is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with Octave; see the file COPYING.  If not, see
<http://www.gnu.org/licenses/>.

*/

#if defined (HAVE_CONFIG_H)
#include <config.h>
#endif

#include <string>
#include <map>
#include <iostream>
#include <algorithm>
#include <cstdlib>

#ifndef OCTAVE_PREFIX
#define OCTAVE_PREFIX %OCTAVE_PREFIX%
#endif

#include "shared-fcns.h"

static std::map<std::string,std::string> vars;

static std::string usage_msg = "usage: octave-config [options]";

static std::string help_msg =
"\n"
"Options:\n"
"\n"
"  -h, -?, --help        Print this message.\n"
"\n"
"  --m-site-dir          Print the name of the directory where Octave\n"
"                        expects to find locally installed .m files.\n"
"\n"
"  --oct-site-dir        Print the name of the directory where Octave\n"
"                        expects to find locally installed .oct files.\n"
"\n"
"  -p VAR, --print VAR   Print the value of the given configuration\n"
"                        variable VAR.  Recognized variables are:\n"
"\n"
"                          API_VERSION            LOCALARCHLIBDIR\n"
"                          ARCHLIBDIR             LOCALFCNFILEDIR\n"
"                          BINDIR                 LOCALOCTFILEDIR\n"
"                          CANONICAL_HOST_TYPE    LOCALSTARTUPFILEDIR\n"
"                          DATADIR                LOCALVERARCHLIBDIR\n"
"                          DATAROOTDIR            LOCALVERFCNFILEDIR\n"
"                          DEFAULT_PAGER          LOCALVEROCTFILEDIR\n"
"                          EXEC_PREFIX            MAN1DIR\n"
"                          EXEEXT                 MAN1EXT\n"
"                          FCNFILEDIR             MANDIR\n"
"                          IMAGEDIR               OCTDATADIR\n"
"                          INCLUDEDIR             OCTFILEDIR\n"
"                          INFODIR                OCTINCLUDEDIR\n"
"                          INFOFILE               OCTLIBDIR\n"
"                          LIBDIR                 PREFIX\n"
"                          LIBEXECDIR             SHLEXT\n"
"                          LOCALAPIARCHLIBDIR     STARTUPFILEDIR\n"
"                          LOCALAPIFCNFILEDIR     VERSION\n"
"                          LOCALAPIOCTFILEDIR\n"
"\n"
"  -v, --version         Print the Octave version number.\n"
"\n";

static void
initialize (void)
{
  vars["OCTAVE_HOME"] = get_octave_home ();
  vars["PREFIX"] = OCTAVE_PREFIX;

  vars["API_VERSION"] = %OCTAVE_API_VERSION%;
  vars["CANONICAL_HOST_TYPE"] = %OCTAVE_CANONICAL_HOST_TYPE%;
  vars["DEFAULT_PAGER"] = %OCTAVE_DEFAULT_PAGER%;
  vars["MAN1EXT"] = %OCTAVE_MAN1EXT%;
  vars["VERSION"] = %OCTAVE_VERSION%;

  vars["ARCHLIBDIR"] = subst_octave_home (%OCTAVE_ARCHLIBDIR%);
  vars["BINDIR"] = subst_octave_home (%OCTAVE_BINDIR%);
  vars["DATADIR"] = subst_octave_home (%OCTAVE_DATADIR%);
  vars["DATAROOTDIR"] = subst_octave_home (%OCTAVE_DATAROOTDIR%);
  vars["EXEC_PREFIX"] = subst_octave_home (%OCTAVE_EXEC_PREFIX%);
  vars["EXEEXT"] = subst_octave_home (%OCTAVE_EXEEXT%);
  vars["FCNFILEDIR"] = subst_octave_home (%OCTAVE_FCNFILEDIR%);
  vars["IMAGEDIR"] = subst_octave_home (%OCTAVE_IMAGEDIR%);
  vars["INCLUDEDIR"] = subst_octave_home (%OCTAVE_INCLUDEDIR%);
  vars["INFODIR"] = subst_octave_home (%OCTAVE_INFODIR%);
  vars["INFOFILE"] = subst_octave_home (%OCTAVE_INFOFILE%);
  vars["LIBDIR"] = subst_octave_home (%OCTAVE_LIBDIR%);
  vars["LIBEXECDIR"] = subst_octave_home (%OCTAVE_LIBEXECDIR%);
  vars["LOCALAPIARCHLIBDIR"] = subst_octave_home (%OCTAVE_LOCALAPIARCHLIBDIR%);
  vars["LOCALAPIFCNFILEDIR"] = subst_octave_home (%OCTAVE_LOCALAPIFCNFILEDIR%);
  vars["LOCALAPIOCTFILEDIR"] = subst_octave_home (%OCTAVE_LOCALAPIOCTFILEDIR%);
  vars["LOCALARCHLIBDIR"] = subst_octave_home (%OCTAVE_LOCALARCHLIBDIR%);
  vars["LOCALFCNFILEDIR"] = subst_octave_home (%OCTAVE_LOCALFCNFILEDIR%);
  vars["LOCALOCTFILEDIR"] = subst_octave_home (%OCTAVE_LOCALOCTFILEDIR%);
  vars["LOCALSTARTUPFILEDIR"] = subst_octave_home (%OCTAVE_LOCALSTARTUPFILEDIR%);
  vars["LOCALVERARCHLIBDIR"] = subst_octave_home (%OCTAVE_LOCALVERARCHLIBDIR%);
  vars["LOCALVERFCNFILEDIR"] = subst_octave_home (%OCTAVE_LOCALVERFCNFILEDIR%);
  vars["LOCALVEROCTFILEDIR"] = subst_octave_home (%OCTAVE_LOCALVEROCTFILEDIR%);
  vars["MAN1DIR"] = subst_octave_home (%OCTAVE_MAN1DIR%);
  vars["MANDIR"] = subst_octave_home (%OCTAVE_MANDIR%);
  vars["OCTDATADIR"] = subst_octave_home (%OCTAVE_OCTDATADIR%);
  vars["OCTFILEDIR"] = subst_octave_home (%OCTAVE_OCTFILEDIR%);
  vars["OCTINCLUDEDIR"] = subst_octave_home (%OCTAVE_OCTINCLUDEDIR%);
  vars["OCTLIBDIR"] = subst_octave_home (%OCTAVE_OCTLIBDIR%);
  vars["SHLEXT"] = subst_octave_home (%OCTAVE_SHLEXT%);
  vars["STARTUPFILEDIR"] = subst_octave_home (%OCTAVE_STARTUPFILEDIR%);
}

int
main (int argc, char **argv)
{
  initialize ();

  if (argc == 1)
    {
      std::cout << usage_msg << std::endl;
      return 1;
    }

  for (int i = 1; i < argc; i++)
    {
      std::string arg (argv[i]);

      if (arg == "-h" || arg == "-?" || arg == "--help")
        {
          std::cout << usage_msg << std::endl;
          std::cout << help_msg;
          return 0;
        }
      else if (arg == "--m-site-dir")
        std::cout << vars["LOCALVERFCNFILEDIR"] << std::endl;
      else if (arg == "--oct-site-dir")
        std::cout << vars["LOCALVEROCTFILEDIR"] << std::endl;
      else if (arg == "-v" || arg == "--version")
        std::cout << vars["VERSION"] << std::endl;
      else if (arg == "-p" || arg == "--print")
        {
          if (i < argc-1)
            {
              arg = argv[++i];
              std::cout << vars[arg] << std::endl;
            }
          else
            {
              std::cerr << "octave-config: " << arg
                        << " options requires argument" << std::endl;
              return 1;
            }
        }
      else
        {
          std::cerr << "octave-config: unrecognized argument " << arg
                    << std::endl;
          return 1;
        }
    }

  return 0;
}
